### PR TITLE
add back copy stageout plugin

### LINKED ISSUE #9144: need for cp stageout plugin
For HPC that do not provide internet access to the worker node, stageout has to happen to the internal file system, which is mounted on the worker nodes.

The CP plugin was retired, so I see two options here. I could write a new CP plugin. Alternatively, the XRDCP plugin understands local file paths for input and output, but it misses the automatic output directory creation. I could add the latter based on parsing the output PFN (only do it if stageout is to a local file system path).

@amaltaro , what do you prefer? Or do you see another option?

### PR DESCRIPTION
**Pull request by hufnagel**
Fixes https://github.com/dmwm/WMCore/issues/9144
Not just the same old Backends/CPImpl, but a rewrite. Added stagein support since this exists in XRDCPImpl, which this new version is based on. Do we even still support stagein? If not maybe we should clean that up, it makes the code much more complicated than it has to be.

The Plugins/CPImpl is similar to the old version, with a few minor tweaks/changes.

@amaltaro , what do you think?

### CODE REVIEW

**Comment by hufnagel** [comment]
Completely untested, just wanted to get some early feedback on code structure etc. Also not sure how I would test this.

**Comment by amaltaro** [comment]
Dirk, I think we can also copy back those 3 test changes. BTW, you have pushed in an empty CPImpl at the root directory as well.

**Comment by hufnagel** [comment]
Added back CPImpl unit test as-is and also adjusted the other unit test. Seems to me we might want to get rid of the special test-copy unit test implementation altogether since it isn't really used now anymore.

**Comment by hufnagel** [comment]
@amaltaro , how can I get cmsdmwmbot to actually give me the output of the unit test?

**Comment by amaltaro** [comment]
Given the list of failures in the report.html, you look them up here:
https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-test/8732/#showFailuresLink

which gives you a link to, for instance:
https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-PR-test/8732/testReport/junit/WMCore_t.Storage_t.Backends_t.CPImpl_t/CPImplTest/testCreateOutputDirectory_simple/

I'm not sure how up-to-date this is, but it might be helpful as well:
https://github.com/dmwm/WMCore/wiki/Understanding-Jenkins

**Comment by hufnagel** [comment]
Thank, that helps, I see what's wrong now. Would be good if the report that is linked in the standard reply would contain that link...

**Comment by hufnagel** [comment]
Ok, this looks good now. @amaltaro , can you review?

Also, what about the other uses of "test-copy"? Should we clean this up?

**Comment by amaltaro** [comment]
The link is made available via "default" check, just push the "Details" link.
If "test-copy" is no longer used anywhere, then I think we can remove it (but I have the impression there are still a few tests using it).

**Comment by amaltaro** [comment]
BTW, you can ask for reviewers at the top right button "Reviewers". It's easier than to find what is waiting for your review.

**Comment by hufnagel** [comment]
I don't think test-copy is actually used anywhere now. There are a few references left though:

[hufnagel@lxplus717 WMCore]$ grep -r test-copy test
test/python/WMCore_t/WMRuntime_t/Scripts_t/SetupCMSSWPset_t.py:            <command value="test-copy"/>
test/python/WMCore_t/WMSpec_t/samples/site-local-config.xml:    <command value="test-copy"/>
test/python/WMCore_t/Misc_t/site-local-config.xml:    <command value="test-copy"/>
[hufnagel@lxplus717 WMCore]$ grep -r test-copy src
src/python/WMCore/Storage/Backends/UnittestImpl.py:registerStageOutImpl("test-copy", LocalCopyImpl)
src/python/WMCore/Storage/Registry.py:                'test-copy': 'TestLocalCopyImpl',

**Comment by hufnagel** [comment]
Ok, requested a review. Even if we want to get rid of these last "test-copy" reference (which at first glance don't seem to be functional to me), that should happen in a separate PR.

File: src/python/WMCore/Storage/Plugins/CPImpl.py
Comment (amaltaro) [inline src/python/WMCore/Storage/Plugins/CPImpl.py]: This `shutil.copy()` doesn't copy all the file metadata. But looking at the history, that's how this plugin used to be, so I think we can leave it like that.

**Review by amaltaro — APPROVED** [review]
Patch looks good, Dirk. Is there any place we can run a few jobs to get it tested before we merge it?

In addition to that, I have two comments:
* when describing your changes in the PR, you can write "Fixes #9144" such that it also closes the issue when it gets merged
* please keep test/* changes and pylint/pep fixes in a different commit than the one with the real src/* changes. No need to change this one, but keep it in mind for future ones please.

**Comment by hufnagel** [comment]
There is no site that uses this as a stageout plugin, that's why I said testing this right now is hard.

**Comment by amaltaro** [comment]
Ok, we shall eventually see whether it works or not. Thanks

**Pull request merged**
